### PR TITLE
Fixed issue where skin material wouldn't render if applied to a model…

### PR DESCRIPTION
**Pull request by antonmic**
… with several submeshes and those submeshes had different material types applied to them.
Also renamed MeshDataInstance to ModelDataInstance.
The fix was to have ModelDataInstance hold an array of object SRGs and assign the correct object SRG to the appropriate submeshes.

Signed-off-by: antonmic <56370189+antonmic@users.noreply.github.com>

### CODE REVIEW

**Review by invertednormal — APPROVED** [review]

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp [hunk 1]
```diff
@@ -215,9 +215,10 @@ namespace AZ
             return {};
         }
 
-        Data::Instance<RPI::ShaderResourceGroup> MeshFeatureProcessor::GetObjectSrg(const MeshHandle& meshHandle) const
+        AZStd::vector<Data::Instance<RPI::ShaderResourceGroup>>& MeshFeatureProcessor::GetObjectSrgs(const MeshHandle& meshHandle) const
         {
-            return meshHandle.IsValid() ? meshHandle->m_shaderResourceGroup : nullptr;
+            static AZStd::vector<Data::Instance<RPI::ShaderResourceGroup>> staticEmptyList;
+            return meshHandle.IsValid() ? meshHandle->m_objectSrgList : staticEmptyList;
```
[the remark this answers is not in the document]
Comment (santorac) [reply Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp:221]: Sorry, I do think it's worth blocking the PR over this. We need to bias toward stable and reliable APIs.

**Review by santorac — CHANGES_REQUESTED** [review]

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp [hunk 2]
```diff
@@ -1148,74 +1154,74 @@ namespace AZ
...
+                    }
                 }
+
+                objectSrg->Compile();
             }
 
-            m_shaderResourceGroup->Compile();
-            m_objectSrgNeedsUpdate = false;
+            // Set m_objectSrgNeedsUpdate to false if there are object SRGs in the list
+            m_objectSrgNeedsUpdate = m_objectSrgNeedsUpdate && (m_objectSrgList.size() == 0);
```
Comment (santorac) [inline Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp:1221]: This logic is a kind of hard to wrap my head around. (Maybe a personal problem :) )
Can't we just set the flag to false like before?

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h [hunk 3]
```diff
@@ -152,7 +153,7 @@ namespace AZ
 
             Data::Instance<RPI::Model> GetModel(const MeshHandle& meshHandle) const override;
             Data::Asset<RPI::ModelAsset> GetModelAsset(const MeshHandle& meshHandle) const override;
-            Data::Instance<RPI::ShaderResourceGroup> GetObjectSrg(const MeshHandle& meshHandle) const override;
+            AZStd::vector<Data::Instance<RPI::ShaderResourceGroup>>& GetObjectSrgs(const MeshHandle& meshHandle) const override;
```
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h:156]: Nope, we actually edit the object SRGs where this function is called to get them (from AtomActorInstance)

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessorInterface.h [hunk 4]
```diff
@@ -66,7 +66,7 @@ namespace AZ
             //! instead of compiling the srg directly. This way, if the srg has already been queued for compile,
             //! it will not be queued twice in the same frame. The ObjectSrg should not be updated during
             //! Simulate, or it will create a race between updating the data and the call to Compile
-            virtual Data::Instance<RPI::ShaderResourceGroup> GetObjectSrg(const MeshHandle& meshHandle) const = 0;
+            virtual AZStd::vector<Data::Instance<RPI::ShaderResourceGroup>>& GetObjectSrgs(const MeshHandle& meshHandle) const = 0;
```
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessorInterface.h:69]: Thanks for catching this, will update

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp (hunk 2 above)
Comment (antonmic) [reply Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp:1221]: This is exactly the same logic as
if(m_objectSrgList.size() > 0)
    m_objectSrgNeedsUpdate = false;
This is to mirror the previous logic where we earlied out if m_shaderResourceGroup was null.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (invertednormal) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h:156]: I don't think that's possible (see my suggestion above)

File: Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (amzn-tommy) [reply Gems/Atom/Feature/Common/Code/Source/Mesh/MeshFeatureProcessor.cpp:221]: I like this suggestion. Agree that it is not blocking.

**Review by amzn-tommy — APPROVED** [review]

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (antonmic) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Mesh/MeshFeatureProcessor.h]: Actually did work, updated the PR

**Review by santorac — APPROVED** [review]

**Pull request merged**